### PR TITLE
Feature: Unconsent permissions

### LINKED ISSUE #1844: Add support for unconsenting a permission
**Is your feature request related to a problem? Please describe.**
It is currently not possible from within Graph Explorer to remove a previously consented scope to test scopes with lower privileges.

- [x] Implement revoke consent
- [x] Enable tenant admins to revoke consent for tenant-wide permissions
- [x] Add a column for consentType
- [x] UI enhancements: Make 'Admin consent required' tab title wrap on small screens, Disable Unconsent button for users with no required permissions
- [x] Add info icon to Unconsent button and show consent-type column regardless of token being present

### PR DESCRIPTION
**Pull request by Onokaev**
## Overview

Closes #1844 

### Demo

Optional. Screenshots, `curl` examples, etc.

### Notes

Optional. Ancillary topics, caveats, alternative strategies that didn't work out, anything else.

## Testing Instructions

* Checkout to this branch

### Test permissions tablist UI

* Open Modify Permissions tab and observe fixes as pointed out on the issue

### Test revoking of permissions

* Try revoking without consenting to the required permissions that enable revoking i.e. `DelegatedPermissionGrant.ReadWrite.All` and `Directory.Read.All` and observe behavior & status message

_Ensure that you have consented to the required permissions to enable consent before trying the following tests_
* Try revoking a default permission e.g. `User.Read` and observe behavior & status message.
* Try revoking to any other permission and observe behavior.

### CODE REVIEW

File: src/app/services/reducers/permissions-reducer.ts [hunk 1]
```diff
@@ -55,3 +56,35 @@ export function scopes(state: IScopes = initialState, action: IAction): any {
...
+    case REVOKE_PERMISSION_ERROR:
+      return {
+        pending: false,
+        error: action.response,
+        data: []
+      }
+    default:
+      return state;
+  }
+}
```
Comment (thewahome) [inline src/app/services/reducers/permissions-reducer.ts:90]: Since we do not store the permissions that we have unconsented to, why not reuse the consenting state objects?

**Review by thewahome — CHANGES_REQUESTED** [review]

File: src/app/views/query-runner/request/permissions/TabList.tsx [hunk 2]
```diff
@@ -55,18 +56,27 @@ const TabList = ({ columns, classes, renderItemColumn, renderDetailsHeader, maxH
...
+        onMouseEnter={() => {
+          // Check if the device width is less than 1260px
+          // OR if the browser width is less than 1290 ie. the browser is resized
+
+          if(screen.width < 1260 || window.innerWidth < 1290){
+            setIsScreenSizeReduced(true);
+          }
+        }
+        }
+        onMouseLeave={() => setIsScreenSizeReduced(false)}>
```
Comment (thewahome) [inline src/app/views/query-runner/request/permissions/TabList.tsx:76]: Is this related to revoking consent?

File: src/types/permissions.ts [hunk 3]
```diff
@@ -41,3 +41,9 @@ export interface IScopes {
   };
   error: any | null;
 }
+
+export interface IUnconsent {
+  pending: boolean;
+  error: any | null;
+  data: string[];
+}
```
Comment (thewahome) [inline src/types/permissions.ts:49]: This signature is reused. It may make sense to rename it to something generic and use the new interface to define any new item that follows the same signature

File: src/app/services/reducers/permissions-reducer.ts (hunk 1 above)
Comment (Onokaev) [reply src/app/services/reducers/permissions-reducer.ts:90]: Yes. We will re-use those

File: src/app/views/query-runner/request/permissions/TabList.tsx (hunk 2 above)
Comment (MaryannGitonga) [reply src/app/views/query-runner/request/permissions/TabList.tsx]: It is part of the ticket. Please refer to #2063

File: src/types/permissions.ts (hunk 3 above)
Comment (MaryannGitonga) [reply src/types/permissions.ts:49]: It is removed by convention when we remove the states in the `src/app/services/reducers/permissions-reducer.ts`.

File: src/app/services/actions/permissions-action-creator.ts [hunk 4]
```diff
@@ -142,3 +148,157 @@ export function consentToScopes(scopes: string[]): Function {
...
+      value: 'no-cache'
+    }
+  ],
+  selectedVersion: '',
+  sampleUrl: ''
+};
+
+const userHasRequiredPermissions = (requiredPermissions: string[],
+  consentedScopes: string[]) => {
+  return requiredPermissions.every(scope => consentedScopes.includes(scope));
```
Comment (thewahome) [inline src/app/services/actions/permissions-action-creator.ts:251]: The functionality here can also be moved to the getPermissionResponse function so that we do not need to import the Graph Client. `getPermissionResponse` can be renamed to something more generic

**Comment by ElinorW** [comment]
Just a heads up...
When I tried to dissent and my browser had blocked pop ups...
I received this error and I had to sign out then sign back in
image

Also, after consenting to a permission and dissenting a different permission. I get this error
image

**Comment by ElinorW** [comment]
Also, for consistency shall we use **dissent** or **unconsent**?
I see **unconsent** used on the buttons but **dissent** used on the error messages.
I think the appropriate grammar is **dissent**... I'm not sure though

**Comment by Onokaev** [comment]
> Just a heads up... When I tried to dissent and my browser had blocked pop ups... I received this error and I had to sign out then sign back in image

Thanks @ElinorW. I was able to reproduce this. I have already pushed a fix for it. Test again and let me know if it's okay on your end

**Comment by Onokaev** [comment]
> Also, for consistency shall we use **dissent** or **unconsent**? I see **unconsent** used on the buttons but **dissent** used on the error messages. I think the appropriate grammar is **dissent**... I'm not sure though

The requirement was to put 'Unconsent'. Let me change the texts on the messages to unconsent also

**Comment by ElinorW** [comment]
When I click on consent and then cancel the operation, this error shows up...
Then when I consent the error doesn't clear uo
image

**Comment by Onokaev** [comment]
> When I click on consent and then cancel the operation, this error shows up... Then when I consent the error doesn't clear up image

There shouldn't be a popup :). You have tested an old version. The new version does not have the popup

**Comment by Onokaev** [comment]
> > > When I click on consent and then cancel the operation, this error shows up... Then when I consent the error doesn't clear up image
> > 
> > 
> > There shouldn't be a popup :). You have tested an old version. The new version does not have the popup
> 
> This is the pop up I am referring to image

I've seen the issue. This appears to also be on prod. Let's log it as a new bug and work on it separately. We are not setting a new status message when we consent to new scopes

File: src/app/services/actions/permissions-action-creator.ts [hunk 5]
```diff
@@ -60,6 +69,39 @@ export function fetchScopesError(response: object): AppAction {
...
+  return {
+    type: REVOKE_SCOPES_SUCCESS
+  }
+}
+
+export function revokeScopesError(): any{
+  return {
+    type: REVOKE_SCOPES_ERROR
+  }
+}
```
Comment (thewahome) [inline src/app/services/actions/permissions-action-creator.ts:103]: Let's have these strongly typed

File: src/app/services/actions/permissions-action-creator.ts [hunk 6]
```diff
@@ -155,3 +197,140 @@ export function consentToScopes(scopes: string[]) {
...
+
+enum REVOKE_STATUS {
+  success = 'success',
+  failure = 'failure',
+  preliminaryChecksFail = 'preliminaryChecksFail',
+  allPrincipalScope = 'allPrincipalScope'
+}
+
+export function revokeScopes(permissionToRevoke: string) {
```
Comment (thewahome) [inline src/app/services/actions/permissions-action-creator.ts:216]: This function looks too long. Is there a way for us to break it down more so that it's easier to read?

**Review by thewahome — APPROVED** [review]
